### ISSUE TITLE
Query about some Training information

### ISSUE THREAD

**Issue by sbmaruf** [facebookresearch/XLM#113]
Hi there!
Excellent work. I have some question.
1. For training Why did you take each of the batches from the same language? Is there any reason for it?
2. For 256 concatenated token stream, did you drop tokens based on sentence or length? Meaning if total tokens are over 256, did you just dropped the tokens or dropped the tokens of the last sentence.
3. From paper section 5.1,

> In our experiments, using either max-pooling or mean-pooling over the last layer did not work better than using the first hidden state

Why the first token `[/s]` represents the entire stream representation?
@glample @aconneau

**Comment by aconneau** [comment]
Hi @sbmaruf,
1) We don't actually, because updates are done on multiple GPUs and each GPU sees a batch from a random language. Since gradients are averaged, the gradient step is made of samples from various languages. In general to answer your question, there is no reason for it no.
2) based on length: there is no padding, the last sentence can be truncated.
3) We did not use a specific token for the stream representation at fine-tuning time, as the positional embedding of the first position can play that role anyway
Hope that helps,
Alexis

**Comment by sbmaruf** [comment]
Hi @aconneau thank you for the quick response.
Sorry to bother you again. I think I have mixed the first hidden state and `[s]` symbol.
Can you elaborate a bit about the following line from [section 5.1](https://arxiv.org/pdf/1901.07291.pdf), 

> In our experiments, using either max-pooling or mean-pooling over the last layer did not work better than using the first hidden state

while reading the paper I was confused what does it mean.

**Comment by aconneau** [comment]
When fine-tuning the pretrained MLM model on downstream tasks, we use the first hidden state of the last layer of the Transformer as an input to the task-specific classifier. Instead of using the first hidden state, we tried averaging all the hidden states of the last layer of the Transformer (mean-pooling) or taking the dimension-wise max (max-pooling) but we did not observe improvements by doing that. So we do use the first hidden state (whose input is indeed the [/s] symbol in our case, and not a [cls] token as in BERT.

**Comment by aconneau** [comment]
Closing the issue now, please feel free to re-open if needed.

**Comment by hughperkins** [comment]
question: in the vocabulary there is both `<s>` and `</s>`. What is `<s>` used for?

**Comment by glample** [comment]
`<s>` is not used anywhere, we always use `</s>` before and after sentences. This way, the model can be trained with a language modeling loss (where sentences are concatenated after each other, i.e. where a single `</s>` separates consecutive sentences, which we call `StreamDataset` in the code) and with a machine translation loss (where you have a `</s>` before and after each sentence), without being confused.

If you have a passage of text with 3 sentences `sent1`, `sent2` and `sent3`, then you will do:
`</s> sent1 </s> sent2 </s> sent3 </s>`. In that case you will have to detect the location of sentences on your own. However, you can also just do: `</s> sent1 sent2 sent3 </s>` and it will probably work equally well.
